src/Utilities/Options/CategorizedAnalyzerConfigOptions.cs

```diff
@@ -0,0 +1,179 @@
+
+            var specificOptions = specificOptionsBuilder.Count > 0 ?
+                specificOptionsBuilder
+                    .Select(kvp => new KeyValuePair<string, ImmutableDictionary<string, string>>(kvp.Key, kvp.Value.ToImmutableDictionary(StringComparer.OrdinalIgnoreCase)))
+                    .ToImmutableDictionary(StringComparer.OrdinalIgnoreCase) :
+                ImmutableDictionary<string, ImmutableDictionary<string, string>>.Empty;
+
+            return new CategorizedAnalyzerConfigOptions(generalOptions, specificOptions);
+        }
+
+        public TEnum GetEnumOptionValue<TEnum>(string optionName, DiagnosticDescriptor rule, TEnum defaultValue)
+            where TEnum : struct
+        {
+            if (TryGetSpecificOptionValue(rule.Id, out var optionValue) ||
+                TryGetSpecificOptionValue(rule.Category, out optionValue) ||
+                TryGetGeneralOptionValue(out optionValue))
+            {
+                return optionValue;
+            }
+
+            return defaultValue;
+
+            // Local functions.
+            bool TryGetSpecificOptionValue(string specificOptionKey, out TEnum specificOptionValue)
+            {
+                if (SpecificOptions.TryGetValue(specificOptionKey, out var specificRuleOptions) &&
+                    specificRuleOptions.TryGetValue(optionName, out var valueString))
+                {
+                    return TryParse(valueString, out specificOptionValue);
+                }
+
+                specificOptionValue = defaultValue;
+                return false;
+            }
+
+            bool TryGetGeneralOptionValue(out TEnum generalOptionValue)
+            {
+                if (GeneralOptions.TryGetValue(optionName, out var valueString))
+                {
+                    return TryParse(valueString, out generalOptionValue);
+                }
+
+                generalOptionValue = defaultValue;
+                return false;
+            }
+
+            bool TryParse(string value, out TEnum result)
+                => Enum.TryParse(value, ignoreCase: true, result: out result);
+        }
+
+        public int GetIntegralOptionValue(string optionName, DiagnosticDescriptor rule, int defaultValue)
+        {
+            if (TryGetSpecificOptionValue(rule.Id, out var optionValue) ||
+                TryGetSpecificOptionValue(rule.Category, out optionValue) ||
+                TryGetGeneralOptionValue(out optionValue))
+            {
+                return optionValue;
+            }
+
+            return defaultValue;
+
+            // Local functions.
+            bool TryGetSpecificOptionValue(string specificOptionKey, out int specificOptionValue)
+            {
+                if (SpecificOptions.TryGetValue(specificOptionKey, out var specificRuleOptions) &&
+                    specificRuleOptions.TryGetValue(optionName, out var valueString))
+                {
+                    return int.TryParse(valueString, out specificOptionValue);
```

Comment:
what does it do for specificRuleOptions?

Reply:
Sorry, didn't understand the question..